ingress: add websockets configuration

nikhiljha opened the pull request:
- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible. -> Does this need to be tested? I feel like this is an Envoy thing.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

Some popular web software (e.x. Jupyterhub, Home Assistant) requires websocket support from the ingress. This commit provides an annotation to enable this support in Envoy. It is disabled by default.

Fixes: #20427

sayboras commented on operator/pkg/ingress/annotations/annotations.go:
```diff
@@ -16,6 +16,7 @@ const (
 	TCPKeepAliveIdleAnnotation             = annotation.Prefix + "/tcp-keep-alive-idle"
 	TCPKeepAliveProbeIntervalAnnotation    = annotation.Prefix + "/tcp-keep-alive-probe-interval"
 	TCPKeepAliveProbeMaxFailuresAnnotation = annotation.Prefix + "/tcp-keep-alive-probe-max-failures"
+	WebsocketsEnabledAnnotation            = annotation.Prefix + "/websockets"
```
nit: websockets -> websocket

sayboras commented on operator/pkg/ingress/annotations/annotations.go:
```diff
@@ -24,9 +25,10 @@ const (
 	defaultTCPKeepAliveInitialIdle   = 10 // in seconds
 	defaultTCPKeepAliveProbeInterval = 5  // in seconds
 	defaultTCPKeepAliveMaxProbeCount = 10
+	defaultWebsocketsEnabled         = 1 // 1 - Enabled, 0 - Disabled
```
the default value could be disabled

nikhiljha replied:
Having it disabled breaks a bunch of end user applications by default. Is it okay to have it enabled so that Cilium ingress works the same as nginx and traefik by default?

nikhiljha replied:
That said, I've gone ahead and made this change anyway.

sayboras replied:
good question, I think it's a mix between different ingress controllers.  AFAIK, contour has WebSocket disabled by default.

sayboras reviewed (commented):
Thanks for your PR, I have below two small comments, let me know if it's making sense.

Will kick off the Ingress Conformance test later.

nikhiljha committed:
ingress: add websockets configuration

Some popular web software (e.x. Jupyterhub, Home Assistant) requires websocket support from the ingress. This commit provides an annotation to enable this support in Envoy. It is enabled by default to be in-line with other popular Ingress implementations (Traefik, ingress-nginx).

Fixes: #20427

Signed-off-by: Nikhil Jha <hi@nikhiljha.com>

sayboras commented on operator/pkg/ingress/envoy_config.go:
```diff
@@ -261,7 +261,7 @@ func getResources(ingress *slim_networkingv1.Ingress, backendServices []*v2.Serv
 
 func getListenerResource(ingress *slim_networkingv1.Ingress, secretNamespace string, enforcedHTTPS bool) (v2.XDSResource, error) {
```
can you add a small unit tests to cover changes in this PR ? 

https://github.com/cilium/cilium/blob/master/operator/pkg/ingress/envoy_config_test.go#L49

sayboras commented on operator/pkg/ingress/envoy_config.go:
```diff
@@ -360,8 +360,16 @@ func getListenerResource(ingress *slim_networkingv1.Ingress, secretNamespace str
 	var connectionManager envoy_extensions_filters_network_http_connection_manager_v3.HttpConnectionManager
+	var upgradeConfigs []*envoy_extensions_filters_network_http_connection_manager_v3.HttpConnectionManager_UpgradeConfig
+
+	if annotations.GetAnnotationWebsocketEnabled(ingress) == 1 {
```
nit: I reckon enabled/disabled will be widely used with the increase of supported annotations in the future, better to have enums/const for the same. Happy to leave it as what we have now as well.

nikhiljha replied:
I think I'll leave it as is for now then :)

sayboras reviewed (commented):
Looks good to me, one small comment on unit tests.

sayboras reviewed (approved):
thanks and lgtm 💯
